### PR TITLE
Optimizes CI builds heavily with selective checks 

### PR DESCRIPTION
**Pull request by potiuk**
* Images are not built if the change is not touching code or docs.
* In case we have no need for CI images we run stripped-down
  pre-commit checks which skip the long checks and only run for
  changed files
* If none of the CLI/Providers/Kubernetes/WWW files changed
  the relevant tests are skipped, unless some of the core files
  changed as well.
* The selective checks logic is explained and documented.

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).

### CODE REVIEW

**Comment by potiuk** [comment]
This is very nice optimisation. It skips all image building part in case the change does not touch .py or .rst files (or any of the scripts/Dockerfiles). Also skips chart tests and api tests at all if charts/apis are not changed. 

This does not happen very often, but if someone just correts typo in one of the .md documents. or .rsts which are not part of the documentation, this will be HUGE decrease of the strain in GA (the job finishes in < 5 minutes)

**Comment by potiuk** [comment]
Hello @ashb @kaxil @turbaszek -> this one is good despite showing "in-progress". It will start working fine only after merging because it modifies the scripts run in "workflow_run". It was running here https://github.com/potiuk/airflow/actions/runs/307602983 and succeeding, so I am pretty-confident it will work :)

**Review by ashb — COMMENTED** [review]
I think this logic needs to be documented/mentioned somewhere in one of our dev docs. CI.rst perhaps?

File: .github/workflows/ci.yml
Comment (ashb) [inline .github/workflows/ci.yml]: Why github.sha here, and not the merge commit sha?

File: .github/workflows/ci.yml
Comment (ashb) [inline .github/workflows/ci.yml:149]: (Not a change on this PR, but why do we care what repository we are for helm chart tests?)

File: .github/workflows/ci.yml
Comment (ashb) [inline .github/workflows/ci.yml]: Running (or not running) static checks based on if we need to build an image or not seems confusing. I know this makes sense right now, but it feels like a side-effect: for instance a possible optimization we might do in future once we have more selective tests of providers is to rebuild the image much less often and mount sources in.

Could we come up with a new name for this output, even if it's always set in the same place as needs-image-build right now?

File: .github/workflows/ci.yml
Comment (ashb) [inline .github/workflows/ci.yml]: Need to change this name.

"No CI" when it's run in CI is... confusing :)

How about static-checks-simple-changes and "Static checks for simple changes only"?

File: scripts/ci/selective_tests.sh
Comment (ashb) [inline scripts/ci/selective_tests.sh:176]: This should include airflow/models/ too I think -- at lot of query logic the API relies on lives in the model code.

File: scripts/ci/selective_tests.sh
Comment (ashb) [inline scripts/ci/selective_tests.sh]: Why have `api_triggering_patterns` local var at all?

```suggestion
    pattern_array=(
        "^airflow/api"
    )
```

File: scripts/ci/selective_tests.sh
Comment (ashb) [inline scripts/ci/selective_tests.sh:203]:
```suggestion
        "\.py$"
        "^CHANGELOG\.txt"
```

File: scripts/ci/selective_tests.sh
Comment (ashb) [inline scripts/ci/selective_tests.sh]: Could this fn not be reimplemented instead to just check that none of the previous selective tests ran anything? Saves having to keep multiple things in sync.

File: scripts/ci/selective_tests.sh
Comment (ashb) [inline scripts/ci/selective_tests.sh]: All of the `show_changes_files` are being passed an argument that isn't even used.
```suggestion
    show_changed_files
```

No reference to `$@` or `$1` etc in 
https://github.com/apache/airflow/blob/3447b55ba57a06c3820d1f754835e7d7f9a1fc68/scripts/ci/selective_tests.sh#L108-L116

File: scripts/ci/selective_tests.sh
Comment (ashb) [inline scripts/ci/selective_tests.sh]: How about.

```suggestion
function run_kubernetes_tests() {
    initialization::ga_output run-kubernetes-tests "${1:-true}"
}
```

And then we can just call it as:

```diff
-if [[ ${kubernetes_tests_needed} == "true" ]]; then
-    run_kubernetes_tests
-else
-    skip_running_kubernetes_tests
-fi
+run_kubernetes_tests "${kubernetes_tests_needed}"
```

Comment (ashb) [reply scripts/ci/selective_tests.sh]: (Or maybe `set_run_kubernetes_tests "${kubernetes_tests_needed}"` might be slightly clearer)

**Review by ashb — DISMISSED** [review]
See individual comments.

This will be great!

File: .github/workflows/ci.yml
Comment (potiuk) [reply .github/workflows/ci.yml]: The ``github.sha`` IS the merge sha for PRs (and for all other types of events we do not use it). It's not quite apparent, GH documentation is not clear about it entirely but you can read about it for example here: https://frontside.com/blog/2020-05-26-github-actions-pull_request/#how-does-pull_request-affect-actionscheckout

File: .github/workflows/ci.yml
Comment (potiuk) [reply .github/workflows/ci.yml:149]: We exclude the workflow from running only for scheduled workflows (hence the || != schedule). Since we are using the CI builds nightly, if someone forks our repository, those scheduled builds (read further) executed also in forks. And several people pointed out that it is rather intrusive thing - to get people use their free jobs from GitHub actions on the forks. Public repos are free, but individuals have limits on a number of jobs and API calls per hour and running scheduled actions on someone else's fork is a bad idea. 

That's why the recommendation (in many threads) was to only run scheduled workflows in the main repository doing this kind of if statement (it basically means the the workflow will run in you fork for pushes and PRs but not for schedules).

This is going to change soon however. It has not been publicly announced by GitHub on their web pages (also the fact that scheduled workflows run in forks was not an obvious thing) but I recently - 1st of October got an email stating that the policy, and we will be able to remove this if on October 27, 2020:

> Updates to GitHub Actions usage policy: scheduled workflows in inactive and forked public repositories.

> Scheduled workflows will be disabled by default in forks of public repos and in public repos with no activity for 60 consecutive days.
We’re making two changes to the usage policy for GitHub Actions. These changes will enable GitHub Actions to scale with the incredible adoption we’ve seen from the GitHub community. Here’s a quick overview:

> Starting today, scheduled workflows will be disabled by default in new forks of public repositories.
Scheduled workflows will be disabled in public repos with no activity for 60 consecutive days.
We’re dedicated to helping the open source community be productive, which is why GitHub Actions is free for the 50 million developers who work with public repositories. These changes will help us ensure a great GitHub Actions experience for the next 50 million developers on GitHub.

> To further evaluate how this will affect you, please read our FAQ below and contact GitHub Support if you run into any problems or need help. You can also always get up to date information about our usage policies in our Docs.

> Thank you for using GitHub Actions 💙

> What will happen to scheduled workflows in forks I already have?
> If you already have scheduled workflows in forks of public repositories, they will be disabled on October 27, 2020. Repository owners will receive a reminder email 7 days prior and can choose to keep the workflows running at that time if they are needed.

> I have inactive public repos, are my scheduled workflows being disabled?
We are not evaluating public repos for 60 consecutive days of inactivity until October 27, 2020. After that, public repos with 60 consecutive days of inactivity will have scheduled workflows disabled. Inactive repo owners will receive a reminder email 7 days before any workflows are disabled and can choose to keep them running at that time.

> Can I re-enable disabled workflows?
Yes! You can easily re-enable any workflow that has been disabled by using the brand new feature to enable & disable workflows. You can enable a disabled workflow with a few clicks.

File: .github/workflows/ci.yml
Comment (potiuk) [reply .github/workflows/ci.yml]: Actually. it's a good name in this case (I even had it done like that before). I thought about renaming it, but I think it reflects pretty well both intention and behaviour.

With this change,, we add two versions of static checks - one that runs with CI image built ("Static checks"  and the other without "Statich checks: no CI image"). If (in selective test scripts) we decide that image build is needed - we run the full set of static checks, if we decide that we do not need it, we run the "no-ci-image one". I think it is really proper logic and not a side effect at all.

File: .github/workflows/ci.yml
Comment (potiuk) [reply .github/workflows/ci.yml]: Note the "no  CI image" (it's not NO CI). I can replace it with "no image" if it is confusing. The logic is that those are all the test that we can run when there is no "CI Image" available. And we decide whether we need to build the image at all in the "selective tests" phase. I can rename "selective tests" to something else  ("Selective CI steps")? to reflect it further but I think the logic is sound:

* in the "selective" part we choose:

   a) whether we need CI image at alll
   b) whether we need to run helm tests
   c) whether we need to run unit tests
   d) whether we need to run API tests
   e) whether we need to run kubernetes tests
   f)  whether we need to run doc builds

And then all those decisions in CI should be based on those (this is currently happening but maybe the names could be adapted to reflect both indention and behaviour. 


I think if just rename "selective tests" into "selective CI steps"  "no CI image' to "no image needed", I think that would be much better and reflect the intentions WDYT?

File: .github/workflows/ci.yml
Comment (ashb) [reply .github/workflows/ci.yml]: Ah cool

File: scripts/ci/selective_tests.sh
Comment (potiuk) [reply scripts/ci/selective_tests.sh:176]: The intention here is, that if any "core/other" file changes (including models), then always all the tests are run (so all API, unit etc. tests should be always run when those change). So there is no need to add them here specifically. If "models" change, then everything else - including providers, cli, etc. are impacted and all tests should be run.

I will double-check if that's the case. 

This selection is only really used, in case ONLY API files change - then ONLY API tests should be run. This work for sure for the regular "unit" API tests, but I will double check if that's the case for the client generation

File: scripts/ci/selective_tests.sh
Comment (potiuk) [reply scripts/ci/selective_tests.sh]: Yeah I think it was a remnant of the earlier version and we do not need it at all any more. those variables were used elsewhere before but indeed they can be removed completely - you are right. Will do.

File: scripts/ci/selective_tests.sh
Comment (potiuk) [reply scripts/ci/selective_tests.sh:203]: will add.

File: scripts/ci/selective_tests.sh
Comment (ashb) [reply scripts/ci/selective_tests.sh:176]: Ah got it - didn't get that from reading the diff but I didn't try to follow all the logic exactly

File: scripts/ci/selective_tests.sh
Comment (potiuk) [reply scripts/ci/selective_tests.sh]: Not really, because we have to handle the logic to check if any "other" files changed. This is especially important to account for future changes - in order to be on the safe side I count the things that i KNOW can be safely tested in isolation and everything else is considered as ("trigger all tests" if they change)

File: scripts/ci/selective_tests.sh
Comment (potiuk) [reply scripts/ci/selective_tests.sh]: Ah yeah. Cleaning it up

File: scripts/ci/selective_tests.sh
Comment (potiuk) [reply scripts/ci/selective_tests.sh]: Right. Will do it. I was doing it at 1am :)

File: scripts/ci/selective_tests.sh
Comment (potiuk) [reply scripts/ci/selective_tests.sh:176]: I will try to describe it in some way that will be easy to read and grasp :). It's difficult though ...

**Comment by potiuk** [comment]
> I think this logic needs to be documented/mentioned somewhere in one of our dev docs. CI.rst perhaps?

Absolutely

File: .github/workflows/ci.yml
Comment (potiuk) [reply .github/workflows/ci.yml]: Actually your proposal was much better when I thought about it. Renaming it to "basic-checks"

File: .github/workflows/ci.yml
Comment (potiuk) [reply .github/workflows/ci.yml]: I decided to change the name similar to your suggestion and described the logic now.

**Comment by potiuk** [comment]
I think all the points are addressed and we are off to a new world of much faster "small" PR builds. I implemented one more optimisation which I think makes perfect sense and will decrease feedback time for some changes to < 1 minute. 

I limited the pre-commits checks in case of "no-code" changes to only run for the incoming changes and not for all files. I think it makes perfect sense, because those "no-code" checks are usually "per-file" so there is very little risk anything will be broken if we do not check all other files. I am going to still test it but I believe in case someone modifies just a README.md, the feedback time will be < 1 minute.

**Comment by potiuk** [comment]
I am still running some tests with it on my fork BTW. so some fixes might come, but feel free to re-review!

**Comment by potiuk** [comment]
WHOA! Readme only change, completed with all the relevant tests here: https://github.com/potiuk/airflow/pull/128 

Completed in 28s  :tada: :tada: :tada: :tada: :tada: :tada: :tada: :tada:

**Comment by potiuk** [comment]
Right @ashb -> i also made some changes to be very explicit about which commits are used where, documented it all and tested it on my fork. Example readme-only build here:  https://github.com/potiuk/airflow/actions/runs/312327864. PR here: https://github.com/potiuk/airflow/pull/128, Total time including queueing : 1m11s, Static checks: 23s.

**Comment by potiuk** [comment]
@ashb Would be nice to merge that one! We have now quite a nu number of really small /typo kind of changes and this one is going to keep the overhead down a lot for those kinds of changes :)

**Comment by ashb** [comment]
I won't be able to review until I'm at a computer on Monday. If someone else had reviewed then dismiss my review/don't wait for me

**Comment by potiuk** [comment]
Not yet @kaxi @turbaszek maybe You'd like to :) ?

File: .github/workflows/build-images-workflow-run.yml
Comment (kaxil) [inline .github/workflows/build-images-workflow-run.yml]: there is a missing quote here

File: CI.rst
Comment (kaxil) [inline CI.rst]:
```suggestion
and it helps to keep down both the usage of jobs in GitHub Actions as well as CI feedback time to
contributors in case of simpler changes.
```

File: .github/workflows/ci.yml
Comment (kaxil) [inline .github/workflows/ci.yml]:
```suggestion
  # Those checks are run if no image needs to be built for checks. This is for simple changes that
```

File: CI.rst
Comment (kaxil) [inline CI.rst]: Will this separation, bring us back to the issue that occurred a few days back where we were fighting for queues with other ASF projects?

Comment (potiuk) [reply CI.rst]: Not at all. First of all I think it was not the root cause (we will start hitting it anyway in weekdays) but most of all I already changed the approach a lot - learning from the "too many jobs". We still separate tests but rather than run them in separate jobs. Right now we run  all test types belonging to the same backend/python version  in a loop - each of the test types in a separate `docker run` - cleaning the docker remnants between each type. 

So if there is a decision to run Core, API and CLI for example, they will run one after the other - each as a separate "docker run" - but they will be run on the same machine. There is a cleanup between each type and we print also the resources available to see if the previous type did not take too much of resources for the next one for whatever reason. This works like that since Wednesday I think. This PR is mainly about also limiting when docker image build is needed at all - there are some cases (like when you change only an .md file) that no CI Image/PROD image are even needed. In this case we skip building and downloading the image and we only run a limited set of "static checks".

I think it is not an often case, but It is often-enough to get it down from 20 minutes to 20 seconds sometimes.

Comment (potiuk) [reply CI.rst]: I've added a short chapter about it.

File: .github/workflows/ci.yml
Comment (kaxil) [inline .github/workflows/ci.yml]: same here

File: .github/workflows/build-images-workflow-run.yml
Comment (kaxil) [inline .github/workflows/build-images-workflow-run.yml:227]:
```suggestion
          if [[ "${EVENT_NAME}" == "pull_request" ]]; then
```

just as best-practise: https://github.com/koalaman/shellcheck/wiki/SC2086

File: .github/workflows/ci.yml
Comment (kaxil) [inline .github/workflows/ci.yml:123]:
```suggestion
          if [[ "${GITHUB_REGISTRY_WAIT_FOR_IMAGE}" == 'true' ]]; then
```

not strictly needed

File: .github/workflows/ci.yml
Comment (kaxil) [inline .github/workflows/ci.yml:134]:
```suggestion
          if [[ "${EVENT_NAME}" == "pull_request" ]]; then
```

File: .github/workflows/ci.yml
Comment (kaxil) [inline .github/workflows/ci.yml]:
```suggestion
  # Those checks are run if no image needs to be built for checks. This is for simple changes that
```

File: CI.rst
Comment (kaxil) [inline CI.rst]:
```suggestion
* API - Tests for the Airflow API (api and api_connexion folders)
* CLI - Tests for the Airflow CLI (cli folder)
```

File: CI.rst
Comment (kaxil) [inline CI.rst]:
```suggestion
* WWW - Tests for the Airflow webserver (www and www_rbac in 1.10 folders)
```

File: CI.rst
Comment (kaxil) [inline CI.rst]:
```suggestion
Even if the types are separated that share the same backend version/python version is run sequentially
on the same CI machine - each of them in a separate docker run command and with additional docker
cleaning between the steps to not fall into the trap of using too many resources in one big test run
but also not to increase the number of jobs per each Pull Request.
```

File: CI.rst
Comment (kaxil) [inline CI.rst]:
```suggestion
1) In case of direct push (so when PR gets merged) or scheduled run, we always run all tests and checks.
   This is in order to make sure that the merge did not miss anything important. The remainder of the logic
   is executed only in case of Pull Requests.

2) We retrieve which files have changed in the incoming Merge Commit (github.sha is a merge commit
   automatically prepared by GitHub in case of Pull Request, so we can retrieve the list of changed
   files from that commit directly).

3) If any of the important, environment files changed (Dockerfile, ci scripts, setup.py, GitHub workflow
   files), then we again run all tests and checks. Those are cases where the logic of the checks changed
   or the environment for the checks changed so we want to make sure to check everything.
```

File: CI.rst
Comment (kaxil) [inline CI.rst]:
```suggestion
7) If any of the relevant source files that trigger the tests have changed at all. Those are airflow
```

All the other `If` are in capital

File: CI.rst
Comment (kaxil) [inline CI.rst]:
```suggestion
9) Quarantined tests are always run when tests are run - we need to run them often to observe how
   often they fail so that we can decide to move them out of quarantine. Details about the
   Quarantined tests are described in `TESTING.rst <TESTING.rst>`_
```

File: CI.rst
Comment (kaxil) [inline CI.rst]:
```suggestion
10) There is a special case of static checks. In case the above logic determines that the CI image
```

File: CI.rst
Comment (kaxil) [inline CI.rst:774]: Good explanation, well done @potiuk 👏

File: scripts/ci/selective_ci_checks.sh
Comment (kaxil) [inline scripts/ci/selective_ci_checks.sh]: Can we change this i.e it is not clear.

For example :

```
# $1 - Merge commit SHA (if it is a PR & if SHA is missing, then all tests are run)
```

???

File: .github/workflows/ci.yml
Comment (potiuk) [reply .github/workflows/ci.yml]: True.

File: .github/workflows/build-images-workflow-run.yml
Comment (potiuk) [reply .github/workflows/build-images-workflow-run.yml:227]: Not really  -  inside the '[[ ]]' the left hand side part of variable does not need quoting and it shellheck does not complain https://unix.stackexchange.com/questions/68694/when-is-double-quoting-necessary 

I actually every were use convention that I omit the quotes everywhere in bash in the left-siide hand of [[ ]]. This makes it a bit less verbose.

File: .github/workflows/ci.yml
Comment (potiuk) [reply .github/workflows/ci.yml:123]: Same here  - there are possibly few hundreds of places all over our code where " no quote on the left side inside [[ ]] " is  followed. I'd prefer to keep it - it makes the code a bit more readable.

File: .github/workflows/ci.yml
Comment (potiuk) [reply .github/workflows/ci.yml:134]: Same here

File: CI.rst
Comment (potiuk) [reply CI.rst:774]: Ahh.. :heart: .. It was REALLY difficult to explain it :)

File: .github/workflows/build-images-workflow-run.yml
Comment (kaxil) [reply .github/workflows/build-images-workflow-run.yml:227]: yea, as I said in another comment, not strictly required but having quotes, is harmless and just for uniformity since we have got quotes in other places ([example](https://github.com/apache/airflow/blob/master/scripts/docker/install_mysql.sh#L21)) :) but yeah feel free to ignore it :) .

File: scripts/ci/selective_ci_checks.sh
Comment (potiuk) [reply scripts/ci/selective_ci_checks.sh]: Updated the description to be clearer. Simply in 'push' and 'schedule' workflow runs we should run all the tests and in this case we call the scripts without SHA (because there is no Merge SHA in this case and we want to run all tests anyway).

**Comment by potiuk** [comment]
All fixed :).

**Review by kaxil — APPROVED** [review]

**Comment by potiuk** [comment]
I've also added a nice (I think) and (missing) for sure - intro at the beginning of the CI.rst - describing some basic assumption and ideas behind the optimizations implemented in our super-complex and some say - overengineered CI. But I think we will see the benefits of it very soon  :)

**Comment by potiuk** [comment]
If any of you around: https://github.com/apache/airflow/pull/11641

cc: @ephraimbuddy 

It turned out that using just merge_commit_sha is not enough. 

The logic of github as understand it now (nowhere to be seen):

1) If you have one commit directly from master -> merge_commit_sha = null, github.sha = your change
2) If you have two commits directly from master or even single one not from master -> merge_commit_sha = merge_commit, github.sha = your latest commit

¯\_(ツ)_/¯

So the logic I have to implement: 

if merge_commit_sha != null -> use merge_commit_sha
else use github.sha 

¯\_(ツ)_/¯

**Pull request merged**
